### DISCUSSION TITLE
Additional music tracks

### DISCUSSION THREAD

**Discussion by Darkloke** [ihhub/fheroes2#9122] [Ideas]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

Greetings. Could you add an ability for user to add his own music tracks as additional alternatives for the current game tracks?

### Describe the possible solution

I suppose all current tracks have their own index for playing (like homm2_17 for Grass theme for example), so you could allow user to place additional tracks as homm2_17_01, homm2_17_02 etc.

Please understand me correctly - i am a big fan of HoMM soundtrack (imho HoMM2 is one the best), but everything gets boring over time, so its good to introduce here some "diversity" - for myself i would add different town themes from next HoMM chapters, add some terrain tracks from MM games and add (a lot!) of alternative combat themes from different fantasy games (King's Bounty for example).

I think this would allow users to expand/create their own musical environment w/o need to switch off ingame music and enabling external music player. :)

**Comment by ihhub** [comment]
Hi @Darkloke , your proposed idea requires the following:
- explanation to players how this could happen
- additional UI element to switch between the original music and expanded
- logic for mapping for each track in the code and within the documentation

Since this idea is far from being complete I am moving this topic into the discussion section.

**Reply by Darkloke** [reply]
- explanation to players how this could happen
Just as i described in the original post - you could add tracks simply under new names, starting from the original track. Like homm2_17 = Grass theme, so new Grass themes would be:
 homm2_17_01
 homm2_17_02
 etc.

- additional UI element to switch between the original music and expanded
No UI is needed. Games just "sees" new tracks and plays them as new ones, based on random selection.

@zenseii 
I dont talk about replacing, i am suggesting to support for playing a new ones - so the game is simply randomly selects from a list available. So lets say now you have a 4 combat themes to play - you can increase this choice up to the (place your number here).

**Reply by zenseii** [reply]
Some new UI element would still be needed because the user should be able to toggle this on or off. So what should it look like, because the original game does not have this feature we need new graphical assets.

There is also the problem of continuous playback. If you play one castle theme up to a certain point and the next time you play a new track and then later you go back to the first track, should the music continue to play from where you ended it the first time? What should happen when switching between towns of the same town type, like 5 Knight castles? Should you play a new track 5 times? This wouldn't sound very good imo.

**Reply by Darkloke** [reply]
I dont understand why any UI is needed for this? Its kind of modding feature - if user does this, then he understand why he does this. By default the game wont provide any additional tracks - only if user manually adds them then he will hear them in the game. If he doesnt want to hear them anymore then he just deletes what he added. I mean this stuff its up to user.

But i suppose 2 config options (for fheroes2.cfg) could be useful for this feature (or may be even for default music) in general:

1. Loop tracks: for example if track ended before combat should the game play a new one or repeat the current.
2. Continuous play: if the game select a track previously played should it continue or play from the beginning.

As for playback "what and when" i suppose here could be a place for imagination... I would personally divide all tracks in the game onto groups and define for each a trigger when the game should randomly choose a track to play.

1. Combat themes: combat start.
2. Town and terrain themes. I see here 2 triggers:

- "Track for a turn", so you are basically "lock" a track for current turn, but on next turn you could hear another track.
- "Selection upon selection", so every time you select a hero (for terrain theme) or open a town screen (town theme) you could hear a different track. "Continuous play" could be nice here, so you wont listen them every time from the beginning.

2. AI theme: end turn.
3. Scenario victory.
4. Main menu.
5. Good/evil campaign.

I admit this idea in general turned out to be a more complex, that i thought from the beginning, but i am asking at least to give a support for combat themes, since this is already working in-game mechanism and we could simply increase a pool of available tracks to choose from.

**Reply by oleg-derevenetz** [reply]
> I dont understand why any UI is needed for this?
> [...]
> But i suppose 2 config options (for fheroes2.cfg) could be useful for this feature (or may be even for default music) in general:
> 
> Loop tracks: for example if track ended before combat should the game play a new one or repeat the current.
Continuous play: if the game select a track previously played should it continue or play from the beginning.

And then an Android user appears asking "I want to have all these settings on Android", what will you tell him? Currently we are making efforts to consider mobile devices as first-class devices, and despite the fact that not all the QoL functionality of HoMM2 in general and fheroes2 in particular (especially hotkey-related one) is currently present on mobile for historical reasons, we periodically implement some new UI parts so that a person without a keyboard can also have some QoL (just like it's done for the "spacebar mechanics", for instance). As for proposals of any new features, they should have full UI support from the very beginning, so that players can use them seamlessly on all supported platforms. Otherwise these proposals will most likely be considered unpromising.

**Comment by zenseii** [comment]
You can already change the music by replacing any track in the music folder with a track with the same filename.

The music folder is either in your fheroes2 installation folder or `%appdata/fheroes2`
